### DISCUSSION TITLE
[Algolia] Extend with translations

### DISCUSSION THREAD

**Discussion by KarolJagodzinski** [saleor/saleor#16122] [Ideas]
The Saleor Algolia app currently does not support search by translations. I would like to initiate a discussion on how this issue could be addressed and developed, particularly for those interested in contributing.

Presently, the index in Algolia is structured as:
`<prefix>.<channel>.<currency>.products`

To incorporate translations, we could either add additional indexes for each language or extend each document with the necessary translations.

This will likely require additional configuration to link specific languages to the relevant channels.

Here are a few potential solutions, each with its own pros and cons:

### Solution 1: One Index Per Language

**Pros:**
- Allows for language-specific optimizations and settings (e.g., query language, decompounding, synonyms, plurals, stop words, query rules) for all languages instead of just the primary language for each region.
- Improved relevance, as mixing languages can introduce noise from cross-language search attributes.
- Enables language-specific merchandising rules, custom rankings, and analytics.

**Cons:**
- AI features may not fully benefit from global events, resulting in varied re-ranking and personalized results across different languages.
- Requires copying merchandising rules across language-specific indexes, potentially increasing workload.
- Multiplies the number of records by the number of languages, complicating the pipeline.
- The total number of indexes would be the product of channel, currency, and language counts.

### Solution 2: One Index for All Languages

This approach would involve using attributes such as `en_name: "Product Name in English"` and `fr_name: "Nom du Produit"` within a single index.

**Pros:**
- Simpler management with a single index.
- AI features can leverage global events for re-ranking and personalized results.
- No need to duplicate merchandising rules across multiple indexes.

**Cons:**
- Potentially reduced relevance due to noise from cross-language search attributes.
- Limited language-specific optimizations and settings.
- Complex query logic to handle multiple language attributes.

### Other: Open for discussion
...

### Conclusion

Each approach offers distinct advantages and challenges. 

For anyone who is interested you can find more on [multilingual search Algolia documentation](https://www.algolia.com/doc/guides/managing-results/optimize-search-results/handling-natural-languages-nlp/how-to/multilingual-search/)

----
#### Side note
Additionally, it's important to address further challenges, such as the need to reindex all products for a specific index if the category translation names have changed.

**Comment by lkostrowski** [comment]
I'm not an Algolia expert but I think we should aim to have separate indexes per language, at least for now. In the future maybe we can support both. My reasoning is mainly staying within Algolia's 10KB limit. If we push translations into a single record, we need to translate everything - not only names, but also attributes. This limit may work for one or two extra languages, but eventually, app will stop working once the limit is exceeded. We can't predict it easily - the majority of products can be small, but some may fail to update, causing bugs hard to find.

Also from frontend perspective I believe its much better.

In terms of AI features and other advanced usages - app will always have its limits and if the business model requires to use custom crafted Algolia connection, this is the moment where custom integration should be added

@krzysztofwolski WDYT?

**Comment by krzysztofwolski** [comment]
Agree with all of the points raised by @lkostrowski and as well I'm fan of separate indexes approach.

> This will likely require additional configuration to link specific languages to the relevant channels.

Would say thats a must - reducing number of channel*language combinations. It will have impact on both Algolia account bill and App infrastructure usage.

> Additionally, it's important to address further challenges, such as the need to reindex all products for a specific index if the category translation names have changed. 

That's my "favorite" edge case. Updating translation for one attribute should trigger updating all of the products using it. This means one person adding a few translations to catalog may trigger massive reindex jobs, probably updating some of the products multiple times in a row. (no matter if single or multiple index approach would be chosen)

**Comment by KarolJagodzinski** [comment]
Thanks for your insights, @krzysztofwolski and @lkostrowski.

Just to confirm, do we agree on the approach of creating separate indices for each language and making the channel configurable with associated languages? The proposed naming convention would be:

`<prefix>.<channel>.<currency>.<language>.products`

**Reply by lkostrowski** [reply]
For me it looks reasonable.

To avoid breaking changes, we should keep legacy index `<prefix>.<channel>.<currency>.products` untouched, but we can make it deprecated

**Reply by krzysztofwolski** [reply]
Dont think we need to deprecate: In the configuration we could provide option "default language prefix". 

Nothing set? Default index uses the old naming.
If the setting is changed, the default index will be renamed.
